File: problems/medium/189.rotate_array.cpp
#include <vector>
using namespace std;
/*

"""  [Problem]
        - Given an integer array nums, rotate the array to the right by k steps, where k is non-negative.

    [Complexity]
        - Time : O(n).
        - Space : O(1), use extra space .
"""
*/

class Solution {
public:
    void reverse(vector<int> & nums, int start, int end){
        while(start < end){
            int dummy = nums[start];
            nums[start] = nums[end];
            nums[end] = dummy;
            start +=1;
            end -=1;
        }
    }
    void rotate(vector<int>& nums, int k) {
        k = k % nums.size();
        if (k <0) k += nums.size();

        reverse(nums, 0, nums.size()-1);
        reverse(nums, 0, k-1);
        reverse(nums, k, nums.size()-1);
    }
};

File: problems/medium/189.rotate_array.py

"""  [Problem]
        - Given an integer array nums, rotate the array to the right by k steps, where k is non-negative.

    [Complexity]
        - Time : O(n).
        - Space : O(1).
"""

class Solution(object):

    def reverse(self, nums, start, end):
        while start < end:
            nums[start], nums[end] = nums[end], nums[start]
            start, end = start+1, end-1
        return nums

    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        k = k % len(nums)
        if k < 0:
            k = k+len(nums)
        
        nums = self.reverse(nums, 0, len(nums)-1)
        nums = self.reverse(nums, 0, k-1)
        nums = self.reverse(nums, k, len(nums)-1)
